add docs for workloadPriority (#1170)
kubernetes-sigs/kueue: site/content/en/docs/concepts/workload_priority_class.md

---
title: "Workload Priority Class"
date: 2023-10-02
weight: 6
description: >

>   A priorityClass which value is utilized by Kueue controller and is independent from pod's priority.

Reviewer: ```suggestion
  A priority class which value is utilized by Kueue controller and is independent from pod's priority.
```

---

> Sometimes developers would like to control workload's priority without affecting pod's priority, like these cases:

Reviewer: Use the second person:
A WorkloadPriorityClass allows you to control the Workload's priority without affecting the pod's priority.

Reviewer: Also link to the documentation for Workload.

- want to prioritize workloads that remain inactive for a specific duration
- want to set a lower priority for development workloads and higher priority for production workloads

A WorkloadPriorityClass is a priorityClass which value is utilized by Kueue controller and 

> is independent from pod's priority.

Reviewer: with the suggestion above, this paragraph becomes unnecessary.

A sample WorkloadPriorityClass looks like the following:

```yaml
apiVersion: kueue.x-k8s.io/v1beta1
kind: WorkloadPriorityClass
metadata:
  name: sample-priority
value: 10000
description: "Sample priority"
```

## How to use WorkloadPriorityClass on Jobs

> The WorkloadPriorityClass is specified through a label `kueue.x-k8s.io/priority-class`.

Reviewer: You can specify the WorkloadPriorityClass by setting the label X

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: sample-job
  labels:
    kueue.x-k8s.io/queue-name: user-queue
    kueue.x-k8s.io/priority-class: sample-priority
spec:
...
```

> The following workload is generated by the yaml above. The `PriorityClassName` field can accept either `PriorityClass` or `WorkloadPriorityClass` name as a value. To distinguish, when using `WorkloadPriorityClass`, a `priorityClassSource` field has the `kueue.x-k8s.io/workloadpriorityclass` value. When using `PriorityClass`, a `priorityClassSource` field has the `scheduling.k8s.io/priorityclass` value.

Reviewer: Kueue generates the following workload for the Job above.

```yaml
apiVersion: kueue.x-k8s.io/v1beta1
kind: Workload
metadata:
  name: job-sample-job-7f173
spec:
  priorityClassSource: kueue.x-k8s.io/workloadpriorityclass
  priorityClassName: sample-priority
  priority: 10000
  queueName: user-queue
...
```

> For other job frameworks, you can set `WorkloadPriorityClass` through a label same for Job.

Reviewer: ```suggestion
For other job frameworks, you can set `WorkloadPriorityClass` using the same label.
```

The Following is an example of `MPIJob`.

```yaml
apiVersion: kubeflow.org/v2beta1
kind: MPIJob
metadata:
  name: pi
  labels:
    kueue.x-k8s.io/queue-name: user-queue
    kueue.x-k8s.io/priority-class: sample-priority
spec:
...
```

## The relationship between pod's priority and workload's priority

> There are three scenarios for creating a workload from a job.

Reviewer: When creating a Workload for a given job, Kueue considers the following scenarios:

1. A job specifies both `WorkloadPriorityClass` and `PriorityClass`
- `WorkloadPriorityClass` is used for the workload's priority.
- `PriorityClass` is used for the pod's priority.
2. A job specifies only `WorkloadPriorityClass`
- `WorkloadPriorityClass` is used for the workload's priority.
- `WorkloadPriorityClass` is not used for pod's priority.
3. A job specifies only `PriorityClass`
- `PriorityClass` is used for the workload's priority and pod's priority.

> In certain job frameworks, there are CRDs that have multiple fields related to pod's priorities within their Spec, or pod priorities are defined in a different location compared to other jobs. In such scenarios, Kueue controller utilizes [`JobWithPriorityClass interface`](https://github.com/kubernetes-sigs/kueue/blob/e162f8508b503d20feb9b31fd0b27d91e58f2c2f/pkg/controller/jobframework/interface.go#L81-L84).  

Reviewer: ```suggestion
In certain job frameworks, there are CRDs that:
- Define multiple pod specs, where each can have their own pod priority, or
- Define the overall pod priority in a dedicated field.

In such scenarios, the integration of the CRD with Kueue utilizes the [`JobWithPriorityClass interface`](https://github.com/kubernetes-sigs/kueue/blob/e162f8508b503d20feb9b31fd0b27d91e58f2c2f/pkg/controller/jobframework/interface.go#L81-L84) to obtain the priority class. You can read the documentation for each job integration to learn how the priority class is obtained.
```

(do we have that information?)

Author: > You can read the documentation for each job integration to learn how the priority class is obtained.
(do we have that information?)

No, I couldn't find a good data source.
Instead, I wrote like this:
You can read the **code** for each job integration to learn how the priority class is obtained.

When defining workload priority, the first step is to check if they have a `WorkloadPriorityClass`. If not, the priority determination process follows the logic defined in the `JobWithPriorityClass interface` for each CRD.

## Where workload's priority is used

> The priority of workloads is utilized in queuing, preemption, and other scheduling processes in Kueue.

Reviewer: ```suggestion
The priority of workloads is used for:
- Sorting the workloads in the ClusterQueues and cohorts.
- Determining whether a workload can preempt others.
```

## Workload's priority values are always mutable

> Workload's `Priority` field is always mutable because it might be useful for the preemption. Thanks to this feature, when there is a workload that has been queued for a while, you can rewrite its priority to execute it with higher priority.  

Reviewer: ```suggestion
The Workload's `priority` field is always mutable.
If a Workload has been pending for a while, you can consider updating its priority to execute it earlier,
based on your own policies. 
```

---

> title: "Workload Priority Class"

Reviewer 2: Oh, I missed the other titles. It makes sense.

        containers:
        - image: gcr.io/k8s-staging-perf-tests/sleep:latest
          name: dummy-job
```

For other job frameworks, you can set `WorkloadPriorityClass` through a label same as Job.
The Following is an example of `MPIJob`.

```yaml
apiVersion: kubeflow.org/v2beta1
kind: MPIJob
metadata:
  name: pi
  labels:
    kueue.x-k8s.io/queue-name: user-queue
    kueue.x-k8s.io/priority-class: sample-priority
spec:
.....
```

## The relationship between pod's priority and workload's priority

There are three scenarios for creating a workload from a job (other job frameworks are same).
1. A job specifies both `workloadPriorityClass` and `PriorityClass`
2. A job specifies only `workloadPriorityClass`
3. A job specifies only `priorityClass`

About #1's scenario, the `workloadPriorityClass` is used for the `workload's priority`. On the other hand, the `priorityClass` is used for the `pod's priority`.  
About #2's scenario, the `workloadPriorityClass` is used for the `workload's priority`. On the other hand, `workloadPriorityClass` is not used for `pod's priority`.  
About #3' scenario,  the `PriorityClass` is used for the `workload's priority` and `pod's priority`. 

> 

Reviewer 3: @tenzen-y  this might be too developer focused and a user cannot do too much about it.

> 

Reviewer 2: I was also concerned about what you say. However, actually, which priorities are used depends on the interface.
For example, we can set three priorities (`.spec.runPolicy.priorityClass`, `.spec.labels.kueue.x-k8s.io/priority-class`, and `.spec.mpiReplicaSecs[*].template.spec.priorityClassName`) on the MPIJob.

> 

Reviewer 2: Or, instead of mentioning the interface, raise MPIJob as a specific case.

@alculquicondor Any better ideas?

> 

Author: Thanks for your comments! Added the explanation related to JobWithPriorityClass

> 

Reviewer: I like how Hiroyuki phrased it :)

## Where workload's priority is used

The priority of workloads is utilized in queuing, preemption, and other scheduling processes in Kueue.

## Workload's priority values are always mutable

> Workload's `Priority` field is always mutable because it might be useful for the preemption.  

Reviewer 3: Is this something relevant for a user?

Author: I think so, I added an example

> Workload's `PriorityClass` and `PriorityClassName` fields are immutable for simplicity.

Reviewer 3: Shold it be like? 

```suggestion
Workload's `PriorityClassSource` and `PriorityClassName` fields are immutable for simplicity.
```

```yaml
apiVersion: kueue.x-k8s.io/v1beta1
kind: Workload
metadata:
  name: job-sample-job-7f173
spec:
  priorityClassSource: kueue.x-k8s.io/workloadpriorityclass
  priorityClassName: sample-priority
  priority: 10000
  queueName: user-queue
...
```

For other job frameworks, you can set `WorkloadPriorityClass` using the same label.
The Following is an example of `MPIJob`.

```yaml
apiVersion: kubeflow.org/v2beta1
kind: MPIJob
metadata:
  name: pi
  labels:
    kueue.x-k8s.io/queue-name: user-queue
    kueue.x-k8s.io/priority-class: sample-priority
spec:
...
```

## The relationship between pod's priority and workload's priority

When creating a `Workload` for a given job, Kueue considers the following scenarios:
1. A job specifies both `WorkloadPriorityClass` and `PriorityClass`
- `WorkloadPriorityClass` is used for the workload's priority.
- `PriorityClass` is used for the pod's priority.
2. A job specifies only `WorkloadPriorityClass`
- `WorkloadPriorityClass` is used for the workload's priority.
- `WorkloadPriorityClass` is not used for pod's priority.
3. A job specifies only `PriorityClass`
- `PriorityClass` is used for the workload's priority and pod's priority.

> In certain job frameworks, there are CRDs that:
> - Define multiple pod specs, where each can have their own pod priority, or
> - Define the overall pod priority in a dedicated field.
> In such scenarios, the integration of the CRD with Kueue utilizes the
> [`JobWithPriorityClass interface`](https://github.com/kubernetes-sigs/kueue/blob/e162f8508b503d20feb9b31fd0b27d91e58f2c2f/pkg/controller/jobframework/interface.go#L81-L84)
> to obtain the priority class. You can read the code for each job integration
> to learn how the priority class is obtained.

Reviewer 3: ```suggestion
In certain job frameworks, there are CRDs that:
- Define multiple pod specs, where each can have their own pod priority, or
- Define the overall pod priority in a dedicated field.
By default kueue will take the PriorityClassName of the first PodSet having one set, However the integration of the CRD with Kueue can implement
[`JobWithPriorityClass interface`](https://github.com/kubernetes-sigs/kueue/blob/e162f8508b503d20feb9b31fd0b27d91e58f2c2f/pkg/controller/jobframework/interface.go#L81-L84)
to change this behavior. You can read the code for each job integration
to learn how the priority class is obtained.
```

## Where workload's priority is used

The priority of workloads is used for:

> - Sorting the workloads in the ClusterQueues and cohorts.

Reviewer 3: ```suggestion
- Sorting the workloads in the ClusterQueues.
```

    kueue.x-k8s.io/priority-class: sample-priority
spec:
...
```

## The relationship between pod's priority and workload's priority

When creating a `Workload` for a given job, Kueue considers the following scenarios:
1. A job specifies both `WorkloadPriorityClass` and `PriorityClass`
- `WorkloadPriorityClass` is used for the workload's priority.
- `PriorityClass` is used for the pod's priority.
2. A job specifies only `WorkloadPriorityClass`
- `WorkloadPriorityClass` is used for the workload's priority.
- `WorkloadPriorityClass` is not used for pod's priority.
3. A job specifies only `PriorityClass`
- `PriorityClass` is used for the workload's priority and pod's priority.

In certain job frameworks, there are CRDs that:
- Define multiple pod specs, where each can have their own pod priority, or
- Define the overall pod priority in a dedicated field.
By default kueue will take the PriorityClassName of the first PodSet having one set,
however the integration of the CRD with Kueue can implement
[`JobWithPriorityClass interface`](https://github.com/kubernetes-sigs/kueue/blob/e162f8508b503d20feb9b31fd0b27d91e58f2c2f/pkg/controller/jobframework/interface.go#L81-L84)

> to change this behavior. You can read the code for each job integration

Reviewer: I guess that's what we have today :(

If you have the time, please update the documentation for MPIJob and other kubeflow CRDs to explain how we obtain the priority class for each.
